src/hooks.rs

```diff
@@ -36,365 +31,555 @@ impl From<i32> for Action {
     }
 }
 
-impl Connection {
-    /// `feature = "hooks"` Register a callback function to be invoked whenever
-    /// a transaction is committed.
-    ///
-    /// The callback returns `true` to rollback.
-    #[inline]
-    pub fn commit_hook<'c, F>(&'c self, hook: Option<F>)
-    where
-        F: FnMut() -> bool + Send + 'c,
-    {
-        self.db.borrow_mut().commit_hook(hook);
-    }
+unsafe fn free_boxed_hook<F>(p: *mut c_void) {
+    drop(Box::from_raw(p as *mut F));
+}
 
-    /// `feature = "hooks"` Register a callback function to be invoked whenever
-    /// a transaction is committed.
-    ///
-    /// The callback returns `true` to rollback.
-    #[inline]
-    pub fn rollback_hook<'c, F>(&'c self, hook: Option<F>)
-    where
-        F: FnMut() + Send + 'c,
-    {
-        self.db.borrow_mut().rollback_hook(hook);
-    }
+#[cfg(feature = "preupdate_hook")]
+mod preupdate_hook {
+    use super::free_boxed_hook;
+    use super::Action;
 
-    /// `feature = "hooks"` Register a callback function to be invoked whenever
-    /// a row is updated, inserted or deleted in a rowid table.
-    ///
-    /// The callback parameters are:
-    ///
-    /// - the type of database update (SQLITE_INSERT, SQLITE_UPDATE or
-    /// SQLITE_DELETE),
-    /// - the name of the database ("main", "temp", ...),
-    /// - the name of the table that is updated,
-    /// - the ROWID of the row that is updated.
-    #[inline]
-    pub fn update_hook<'c, F>(&'c self, hook: Option<F>)
-    where
-        F: FnMut(Action, &str, &str, i64) + Send + 'c,
-    {
-        self.db.borrow_mut().update_hook(hook);
+    use std::os::raw::{c_char, c_int, c_void};
+    use std::panic::catch_unwind;
+    use std::ptr;
+
+    use crate::ffi;
+    use crate::types::ValueRef;
+    use crate::{Connection, InnerConnection};
+
+    // TODO: how to allow user access to these functions, since they should be only accessible in
+    // the scope of a preupdate_hook callback.
+    pub struct PreUpdateHookFunctions {
+        db: *mut ffi::sqlite3,
     }
 
-    /// `feature = "hooks"` Register a query progress callback.
-    ///
-    /// The parameter `num_ops` is the approximate number of virtual machine
-    /// instructions that are evaluated between successive invocations of the
-    /// `handler`. If `num_ops` is less than one then the progress handler
-    /// is disabled.
-    ///
-    /// If the progress callback returns `true`, the operation is interrupted.
-    pub fn progress_handler<F>(&self, num_ops: c_int, handler: Option<F>)
-    where
-        F: FnMut() -> bool + Send + RefUnwindSafe + 'static,
-    {
-        self.db.borrow_mut().progress_handler(num_ops, handler);
+    impl PreUpdateHookFunctions {
```

Comment:
This is where I'm not 100% sure how to proceed. These functions are accessible based on the type of action (insert, delete, or update), during the preupdate_hook callback. There are multiple ways to resolve this:

- Pass Option<Fn> to the callback depending on the function
- Pass an Enum for which the value is based on the Action, and the object inside that Enum implements the relevant functions (similar to what is happening now only with an added layer and removal of the unsafe on the function itself.
- Keep it like this and hold the end user responsible for correctly using the functions.